tracer/test/Datadog.Trace.ClrProfiler.Managed.Tests/AutoInstrumentation/AWS/SnsCommonTests.cs

```diff
@@ -0,0 +1,74 @@
+// <copyright file="SnsCommonTests.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using Amazon.SimpleNotificationService.Model;
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.SNS;
+using Datadog.Trace.HttpOverStreams;
+using Datadog.Trace.Propagators;
+using Datadog.Trace.Vendors.Newtonsoft.Json.Linq;
+using FluentAssertions;
+using Moq;
+using Xunit;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.SNSTests
+{
+    public class SnsCommonTests
+    {
+        private const string DatadogAttributeKey = "_datadog";
+
+        private readonly SpanContext _spanContext;
+        private readonly TraceId _traceId;
+        private readonly string _parentId;
+
+        public SnsCommonTests()
+        {
+            ulong upper = 1234567890123456789;
+            ulong lower = 9876543210987654321;
+
+            var newTraceId = new TraceId(upper, lower);
+            ulong spanId = 6766950223540265769;
+            _spanContext = new SpanContext(newTraceId, spanId, 0, "test-service", "origin");
+            _traceId = newTraceId;
+            _parentId = spanId.ToString();
+        }
+
+        [Fact]
+        public void InjectHeadersIntoMessage_AddsDatadogAttribute()
+        {
+            // Arrange
+            var request = new PublishRequest();
+            var proxy = new PublishRequestProxy(request);
+
+            // Act
+            ContextPropagation.InjectHeadersIntoMessage<PublishRequest>(proxy, _spanContext);
+
+            // Assert
+            var attributes = (Dictionary<string, MessageAttributeValue>)proxy.MessageAttributes;
+
+            attributes.Should().ContainKey(DatadogAttributeKey);
+            if (attributes.TryGetValue(DatadogAttributeKey, out var attributeValue))
+            {
+                var ddTraceContextMemoryStream = attributeValue.BinaryValue;
+                ddTraceContextMemoryStream.Position = 0; // Reset the position, in case it's at the end
+                var reader = new StreamReader(ddTraceContextMemoryStream);
+                var jsonString = reader.ReadToEnd();
+
+                var traceContextJson = JObject.Parse(jsonString);
+                var traceId = traceContextJson["x-datadog-trace-id"].Value<string>();
+                var parentId = traceContextJson["x-datadog-parent-id"].Value<string>();
+                Assert.Equal(_parentId, parentId);
+                Assert.Equal("9876543210987654321", traceId);
```

Comment:
@lucaspimentel I wasn't able to grok the TraceId stuff. I tried calling .ToString() on _traceId but it would give me what seems like a hex value `112210f47de98115891087b8e3b70cb1` but traceId (line 63) would give me `9876543210987654321` i.e. the (64 bit) ulong lower defined on line 32.

A mess of bits :)

Reply:
I cleaned up the edits to the Inject method though

Reply:
Historically all trace ids have been `ulong` (unsigned 64-bit integers). Recently we added the `TraceId` type which supports 128-bit trace ids. The tracer still generates 64-bit trace ids by default for backwards compat (the plan is to enable this by default someday). However, we always support 128-bit trace id received from an upstream service via propagated.

`TraceId.ToString()` will always return a hexadecimal string with length 32. If it was a 64-bit id, the string will be padded with zeros.

`ulong.ToString()` returns a decimal string, which is how Datadog has always represented trace ids in the part.

Note that span ids are always of type `ulong`, not `TraceId`.

Reply:
In your code, you're using 128-bit trace id, and decimal notation to set the upper and lower 64 bits:
```csharp
ulong upper = 1234567890123456789;
ulong lower = 9876543210987654321;
```
This is the equivalent of (using hexadecimal notation):
```csharp
ulong upper = 0x112210f47de98115;
ulong lower = 0x891087b8e3b70cb1;
```
when you put those two together in `TraceId.ToString()`, you get
```
112210f47de98115891087b8e3b70cb1
```
which is correct.

However, for backwards compat (again), we keep injecting only the lower 64 bits into the `x-hdatadog-trace-id` header in decimal notation. The upper 64 bits are added as a "propagated tag" in `x-datadog-tags`. That header should have a hex value like `tid=112210f47de98115` (ote this header can contain additional tags like `dm`).

Reply:
For this test, it may be better to use `DatadogContextPropagator.TryExtract()` to create a new `SpanContext` from the headers, and compare those values to the original `SpanContext`.